inference-engine/src/mkldnn_plugin/nodes/mkldnn_topk_node.cpp

```diff
@@ -40,190 +1041,541 @@ bool MKLDNNTopKNode::isSupportedOperation(const std::shared_ptr<ngraph::Node>& o
+        jcp.axis_dim = axis_dim;
+        jcp.mode_max = mode_max;
+        jcp.sort_index = sort_index;
+        jcp.topk_innermost = topk_innermost;
+        jcp.algorithm = algorithm;
+        jcp.sort_stride = static_cast<int>(I);
+        jcp.work_amout = static_cast<int>(I);
+        if (layout == TopKLayoutType::topk_blocked && topk_innermost) {
+            jcp.blk_stride = D * H * W * blk_size;
+            if (algorithm == TopKAlgorithm::topk_bubble_sort) {
+                jcp.work_amout = static_cast<int>(axis_dim);
+            }
+        }
+
+        if (algorithm == TopKAlgorithm::topk_bitonic_sort) {
+            ceil_pow_2(axis_dim, jcp.bitonic_size, jcp.bitonic_idx_cnt, true);
+            if (sort_index) {
+                ceil_pow_2(top_k, jcp.bitonic_k_size, jcp.bitonic_k_idx_cnt, false);
+            }
+        }
+
+        if (mayiuse(cpu::x64::avx512_common)) {
+            topk_kernel.reset(new jit_uni_topk_kernel_f32<cpu::x64::avx512_common>(jcp));
+        } else if (mayiuse(cpu::x64::avx2)) {
+            topk_kernel.reset(new jit_uni_topk_kernel_f32<cpu::x64::avx2>(jcp));
+        } else if (mayiuse(cpu::x64::sse41)) {
+            topk_kernel.reset(new jit_uni_topk_kernel_f32<cpu::x64::sse41>(jcp));
+        }
+
+        if (topk_kernel)
+            topk_kernel->create_ker();
+    } else { //reference mode
+        int j;
+        for (j = src_dims.size() - 1; j >= 0; j--) {
+            if (src_dims[j] != 1)
+                break;
+        }
+        if (static_cast<size_t>(j) == axis)
+            is_last_dim = true;
+        dim = static_cast<int>(src_dims[axis]);
+        before_num = count(src_dims, 0, axis);
+    }
 }
 
 void MKLDNNTopKNode::execute(mkldnn::stream strm) {
-    const float *src = reinterpret_cast<const float *>(getParentEdgeAt(TOPK_DATA)->getMemoryPtr()->GetPtr());
-    src_k = reinterpret_cast<int *>(getParentEdgeAt(TOPK_K)->getMemoryPtr()->GetPtr())[0];
-    float* dst_data = nullptr;
-    int* dst_idx = nullptr;
-
-    if (outDims.size() == 1) {
-        if (getOriginalOutputPrecisionAtPort(0) == Precision::FP32) {
-            dst_data = reinterpret_cast<float *>(getChildEdgesAtPort(0)[0]->getMemoryPtr()->GetPtr());
+    auto &srcMemPtr = getParentEdgeAt(TOPK_DATA)->getMemoryPtr();
+    auto &dstMemPtr = getChildEdgeAt(TOPK_DATA)->getMemoryPtr();
+    auto &dstIndexesMemPtr = getChildEdgeAt(TOPK_INDEX)->getMemoryPtr();
+
+    const uint8_t *src_data = reinterpret_cast<const uint8_t *>(srcMemPtr->GetPtr());
+    uint8_t *dst_data = reinterpret_cast<uint8_t *>(dstMemPtr->GetPtr());
+    uint8_t *dst_idx = reinterpret_cast<uint8_t *>(dstIndexesMemPtr->GetPtr());
+
+    if (jit_mode) {
+        uint8_t *process_data = NULL;
+        uint8_t *process_idx = NULL;
+        if (algorithm == TopKAlgorithm::topk_bitonic_sort) {
+            size_t src_count = srcMemPtr->GetElementsCount();
+            process_data = reinterpret_cast<uint8_t *>(std::malloc(src_count * data_size));
```

Comment:
Two questions.
1) Why not std::vector?
2) Why cannot we allocate memory on createPrimitve stage in order not to affect inference time?

Reply:
Fixed. Thanks Dmitry!